### PR TITLE
Iterable support

### LINKED ISSUE #45: functions working on collections like first should work on iterables
As of now this is not possible making thinks like getting the first value of an iterator not possible.
the using the keyword iterable since version 7.1 should solve this as both primitive arrays and Iterator implementation would are considered iterable . 
https://secure.php.net/manual/en/language.types.iterable.php

### PR DESCRIPTION
**Pull request by MonsieurV**
Accept [Iterables](https://www.php.net/manual/en/language.types.iterable.php) on the collection functions instead of just arrays.

Closes #45

### CODE REVIEW

**Comment by MonsieurV** [comment]
@maciejczyzewski can you invite @allejo as contributor on the project so he can work directly on this PR/branch with me?

**Comment by MonsieurV** [comment]
Ok, a first batch of functions with Iterable support has been implemented.

Notes:
* most of the functions rely on `foreach` or on functions that are based on `foreach` (like `doForEach`, `reduce`), so the code does not change so much;
* `__::reverseIterable` is now exposed as a proper function: it is useful to work on arrays or iterators; (Should we name it simply `__::reverse`? As anyway all our collection functions now should support Iterables and it does just that on simple arrays)
* as discussed in #45, for the moment we support Iterable as inputs, but does not implement functions as feed-oriented (yielding values for functions like `filter` or `map`): this could be another PR as it may break the current API

We still miss the following functions to be tested/checked/reimplemented:
> assign, concat, concatDeep, doForEach, doForEachRight, ease, every, get, groupBy, has, hasKeys, isEmpty, last

Some of them may be tricky (`ease`, `groupBy`, `has`), so do not hesitate to reach out @allejo so we'll check solutions together.

**Comment by allejo** [comment]
@MonsieurV so sorry I went MIA after having committed to working on these functions; I'm back though and I'll be working on this over my holiday break.

Before I work on these, I'd like to run an idea by you. One of the concerns about iterators is forcing reading an entire iterator into memory. Can we just leave this decision to the user? If they want to give our functions their iterators, they should call `iterator_to_array` first. Since arrays are pass by value, most all array functions will typically return a new array, however, objects are pass by reference. Since iterators are objects, what are your thoughts on working directly on the iterators?

As an example, for `drop()` which drops X number items from the beginning of an array, when you give it an iterator it would just advance X values in the iterator. e.g. https://github.com/allejo/bottomline/commit/30dd909be345215812796981aaf3e9293c49ce6e

Following the same principle, for `filter()` we can just move the pointer of the iterator and yield the next value that meets the specified criteria.

And speaking of `yield`, is PHP 5.4 support still necessary that we should be limiting ourselves by needing workarounds for the `yield` keyword?

**Comment by MonsieurV** [comment]
> If they want to give our functions their iterators, they should call `iterator_to_array` first.

I think I agree on that. If at the end our implementation loads the entire iterator in memory, there isn't much value added compared to just calling `iterator_to_array` before using the library.
In the other end if we choose to return iterators when it [makes sense](https://github.com/maciejczyzewski/bottomline/issues/45#issuecomment-536732684), the plus-value is greater: the user can use the library to apply `map`, `filter`, etc. operations on huge data set or data flux without having to load them all in memory.

> Since iterators are objects, what are your thoughts on working directly on the iterators?
 
With `foreach` we work on the iterators. But I think you suggest we work on the passed iterators *and* return iterators (with `yield`) when appropriate; isn't?

> As an example, for `drop()` which drops X number items from the beginning of an array, when you give it an iterator it would just advance X values in the iterator. e.g. [allejo@30dd909](https://github.com/allejo/bottomline/commit/30dd909be345215812796981aaf3e9293c49ce6e)

Yep! However shouldn't our `drop` function says it return the dropped elements?
That's a bit weird to consume on a drop operation.

> Following the same principle, for `filter()` we can just move the pointer of the iterator and yield the next value that meets the specified criteria.

Totally agree. If we can support `yield`, this is the best usage.

> And speaking of `yield`, is PHP 5.4 support still necessary that we should be limiting ourselves by needing workarounds for the `yield` keyword?

We can indeed consider that the library users are more inclined to run decently recent versions of PHP. `yield` is supported since [PHP 5.5 ](https://www.php.net/manual/en/language.generators.overview.php), which is "old enough" for me. So agree to drop PHP 5.4.

**Comment by allejo** [comment]
> > If they want to give our functions their iterators, they should call `iterator_to_array` first.
> 
> I think I agree on that. If at the end our implementation loads the entire iterator in memory, there isn't much value added compared to just calling `iterator_to_array` before using the library.
> In the other end if we choose to return iterators when it [makes sense](https://github.com/maciejczyzewski/bottomline/issues/45#issuecomment-536732684), the plus-value is greater: the user can use the library to apply `map`, `filter`, etc. operations on huge data set or data flux without having to load them all in memory.

I agree, with some functions like `map` and `filter`, it makes sense to have perform `yield` operations.

> > Since iterators are objects, what are your thoughts on working directly on the iterators?
> 
> With `foreach` we work on the iterators. But I think you suggest we work on the passed iterators _and_ return iterators (with `yield`) when appropriate; isn't?

It depends on the function. Like `map` and `filter`, we'd yield the current iterator value. But for something like `drop()`, I'd want to explore to see if it's possible to change the iterator's current position and return the iterator with its internal position changed.

> > As an example, for `drop()` which drops X number items from the beginning of an array, when you give it an iterator it would just advance X values in the iterator. e.g. [allejo@30dd909](https://github.com/allejo/bottomline/commit/30dd909be345215812796981aaf3e9293c49ce6e)
> 
> Yep! However shouldn't our `drop` function says it return the dropped elements?
> That's a bit weird to consume on a drop operation.

See https://github.com/allejo/bottomline/commit/30dd909be345215812796981aaf3e9293c49ce6e#r36541936

> > And speaking of `yield`, is PHP 5.4 support still necessary that we should be limiting ourselves by needing workarounds for the `yield` keyword?
> 
> We can indeed consider that the library users are more inclined to run decently recent versions of PHP. `yield` is supported since [PHP 5.5 ](https://www.php.net/manual/en/language.generators.overview.php), which is "old enough" for me. So agree to drop PHP 5.4.

:tada: 🥳

**Comment by allejo** [comment]
@MonsieurV your thoughts on this attempt for `drop`? https://github.com/maciejczyzewski/bottomline/commit/46b4f0b76897a056120d900c703aebff70484717

I bumped the min requirement to 5.5. Turns out that for iterators, we can manipulate the internal pointer for it, however, if that iterator is fed to a `foreach`, the `foreach` will call a `rewind()` on the iterator so any manipulation we do would be useless. Instead of changing the internal pointer, I'm now returning a `Generator` with `yield`.

However, when using `yield` inside a function, it will create a `Generator`. Meaning if we have this:

```php
function drop()
{
  if (is_array(...)) {
    return [];
  }

  // ...
  yield ...;
}
```

Even if given an array, it will return a generator. To get around this, I've moved the iterator logic to a separate function and return that function instead allowing us to return an array or a generator depending on its input.

**Comment by MonsieurV** [comment]
> @MonsieurV your thoughts on this attempt for `drop`? [allejo@46b4f0b](https://github.com/allejo/bottomline/commit/46b4f0b76897a056120d900c703aebff70484717)

Good for me!
Some comments on the commit, that we can confirm/infirm with more tests.

> Even if given an array, it will return a generator. To get around this, I've moved the iterator logic to a separate function and return that function instead allowing us to return an array or a generator depending on its input.

Indeed there seems to be [no way](https://www.php.net/manual/en/language.generators.syntax.php) around that. This will make the implementation a little more bulky (with each time the array/return vs. iterator/yield implementation), but at least the distinction will be clearer.

**Comment by allejo** [comment]
Alrighty, here's `__::drop()` with your comments applied and a test case for generators added! https://github.com/allejo/bottomline/commit/7fd43277d36827ea470f7a21110990d2b2943600

**Comment by MonsieurV** [comment]
> Alrighty, here's `__::drop()` with your comments applied and a test case for generators added! [allejo@7fd4327](https://github.com/allejo/bottomline/commit/7fd43277d36827ea470f7a21110990d2b2943600)

Neat!

I've tried an implementation of `filter` following your model. Tell me if that's good for you.
For me it feels like we could go on and try to implement all the relevant functions with Generator/yield support.

PS. Great the `dataProvider`s in the tests: way more concise and to the point.

**Comment by allejo** [comment]
> I've tried an implementation of `filter` following your model. Tell me if that's good for you.
> For me it feels like we could go on and try to implement all the relevant functions with Generator/yield support.

Yup! That's what I had envisioned for `filter` 👍 (I'd be in favor of introducing a `$preserveKeys` to `filter` btw)

I'm satisfied with adopting generators/yield for all of the relevant functions, it'll let us be consistent across all our functions.

**Comment by allejo** [comment]
Once this finished, we should probably move our helper functions/classes in our unit tests to a single location. Maybe a folder `tests/internal/` or something along those lines?

**Comment by MonsieurV** [comment]
> Once this finished, we should probably move our helper functions/classes in our unit tests to a single location. Maybe a folder `tests/internal/` or something along those lines?

Yep!
You can do it asap on your branch, I'll forward it here.

**Comment by allejo** [comment]
@MonsieurV for `flatten`, do we want to exhaust the iterator and read it all into an array? Or should we use a generator?

Same question for `map`, `mapKeys`, and `mapValues`.

Also, here's a table of all the functions and their notes because I needed to visualize the progress.

| Function | Iterable Support | Notes |
| ------ | :----: | ---- |
| append | :x: | An iterator can't be added to and to build a generator to return a single extra value doesn't really seem worth it. We'd have to introduce our own iterator type to be able to simulate merging two iterators. |
| assign | :x: | Iterators are read-only, we can't really assign things to its elements |
| chunk | :white_check_mark: |  |
| compact | :white_check_mark: | |
| concat | :x: | Iterators are read-only, we'd have to read seperate iterators into memory; this would be no different than using `iterator_to_array` |
| concatDeep | :x: | See `concat` |
| doForEach | :white_check_mark: | |
| doForEachRight | :white_check_mark: | |
| drop | :white_check_mark: | |
| ease | :construction: | TO DO |
| every | :white_check_mark: | |
| filter | :white_check_mark: | |
| first | :white_check_mark: | |
| flatten | :question: | ... |
| get | :x: | It's possible we'll read the entire iterator into memory; just use `iterator_to_array` |
| groupBy | :construction: | TO DO |
| has | :x: | See `get()` |
| hasKeys | :x: | See `has()` |
| isEmpty | :construction: | TO DO |
| last | :construction: | TO DO |
| map | :warning: | Implementation may change; should this be updated to be a generator? |
| mapKeys | :warning: | See `map()` |
| mapValues | :warning: | See `map()` |
| max | :white_check_mark: | |
| merge | :white_check_mark: | |
| min | :white_check_mark: | |
| patch | :x: | Iterators are read-only, we'd potentially read whole iterator into memory |
| pick | :white_check_mark: | |
| pluck | :white_check_mark: | |
| prepend | :x: | See `append()` |
| randomize | :x: | This would require us to iterate through the iterator twice; just use `iterator_to_array` |
| reduce | :white_check_mark: | |
| reduceRight | :white_check_mark: | |
| reverseIterable | :white_check_mark: | |
| set | :x: | Iterators are read-only |
| unease | :white_check_mark: | |
| where | :white_check_mark: | |

**Comment by allejo** [comment]
Alrighty @MonsieurV, pending my question about the `map` and friends functions, I think this PR is ready for any other testing/cleaning we need to do.

How do you feel about having the functions be standardized to the following policy: if the function needs to be aware of the full iterator's data then we'll return an array. Otherwise, we return a generator. So that'd mean that `map` and friends will be updated to return a generator since it doesn't need to know about the full iterator, just each individual value.

File: src/__/arrays/flatten.php [hunk 1]
```diff
@@ -19,26 +44,20 @@
...
-            foreach ($value as $valItem) {
-                $output[] = $valItem;
-            }
-        } else {
-            $output[] = $value;
-        }
+    $generator = flattenIterable($iterable, $shallow);
+
+    if (is_array($iterable)) {
+        return iterator_to_array($generator);
```
Comment (MonsieurV) [inline src/__/arrays/flatten.php]: Like the pattern (common implementation based on iterable + `iterator_to_array`) :+1:

File: src/__/collections/filter.php [hunk 2]
```diff
@@ -28,14 +49,31 @@
...
+ *                          Otherwise, an array will be returned.
  */
-function filter(array $array, \Closure $closure = null)
+function filter($iterable, \Closure $closure = null)
 {
-    return \array_values(
-        $closure ? \array_filter($array, $closure) : \array_filter($array)
-    );
+    if (is_array($iterable)) {
+        $values = [];
```
Comment (MonsieurV) [inline src/__/collections/filter.php]: Could be factorized with `iterator_to_array(filterIterable($iterable, $closure))`?

**Comment by MonsieurV** [comment]
Sorry @allejo, I totally went MIA - I was taken this January by business.

> for flatten, do we want to exhaust the iterator and read it all into an array? Or should we use a generator?

Definitively a generator (especially for map & related).

> How do you feel about having the functions be standardized to the following policy: if the function needs to be aware of the full iterator's data then we'll return an array. Otherwise, we return a generator.

Feel right!
What you've done (eg. _`__::map()` to maybe return a Generator_) feels actually even better:
* when provided an iterator it returns a generator
* for array or stdClass the existing API is preserved (returns an array).

This is great work @allejo. I think it will make the library potentially pertinent to use with async / non-blocking PHP libraries (see for eg. [ReactPHP](https://github.com/reactphp/reactphp) or [Amp](https://amphp.org/)).

File: src/__/collections/filter.php (hunk 2 above)
Comment (allejo) [reply src/__/collections/filter.php]: Yup! Good call, I missed this one :sweat_smile:

**Comment by allejo** [comment]
Welcome back @MonsieurV! No worries about going MIA, I did too a few months back :sweat_smile:

File: src/__/collections/filter.php (hunk 2 above)
Comment (allejo) [reply src/__/collections/filter.php]: Done in 94d6cf8ee04e3fd9c5b849ab1b70d3a6a976a98b

**Comment by MonsieurV** [comment]
@maciejczyzewski maybe we could have a release? :pray: 

See [CHANGELOG](https://github.com/maciejczyzewski/bottomline/blob/master/CHANGELOG.md)

**Comment by allejo** [comment]
@maciejczyzewski before you tag a new release, could you look at #53?

**Pull request merged**